### ISSUE TITLE
Layer group support

### ISSUE THREAD

**Issue by atztogo** [spglib/spglib#200]
Discussion about layer group support is raised at #199.
Let's continue the discussion here.

### Possible issues
- Mapping between layer group and space group.
- Hard coded `periodic_axes`.

### Test and refactoring

By @lan496,
> For testing, we may need another repository to check layer-group implementations. For example, I've created and checked magnetic-space-group implementations for MAGNDATA in a private repo. For refactoring, wrapper functions to hide comparison of `cell->aperiodic_axis` may be helpful, like `inline int is_aperiodic_axis(const *Cell cell, const int axis).`

### Future plan
Proposed by @site-g
- [ ] Systematic test (Almost done by @mikaelkuisma1)
- [ ] Something in datasets to indicate group type (now negative hall number means layer group)
- [ ] C-APIs (now only `spg_get_layer_dataset`)
- [ ] Python, _etc._ interfaces
- [ ] Documentation for layer group
- [ ] Add `group_class` to `SpglibDataset`

**Comment by blokhin** [comment]
Below is a simple space-layer group mapping logic which however does not account for the lots of the edge cases. It just worked for me ™.

```python
if group_num_3d in [25, 26, 28, 51]:
    warnings.warn('Diperiodical group setting is undefined')

DIPERIODIC_MAPPING = {3:8, 4:9, 5:10, 6:11, 7:12, 8:13, 10:14, 11:15, 12:16, 13:17, 14:18, 16:19, 17:20, 18:21, 21:22, 25:23, 25:24, 26:25, 26:26, 27:27, 28:28, 28:29, 29:30, 30:31, 31:32, 32:33, 35:34, 38:35, 39:36, 47:37, 49:38, 50:39, 51:40, 51:41, 53:42, 54:43, 55:44, 57:45, 59:46, 65:47, 67:48, 75:49, 81:50, 83:51, 85:52, 89:53, 90:54, 99:55, 100:56, 111:57, 113:58, 115:59, 117:60, 123:61, 125:62, 127:63, 129:64, 143:65, 147:66, 149:67, 150:68, 156:69, 157:70, 162:71, 164:72, 168:73, 174:74, 175:75, 177:76, 183:77, 187:78, 189:79, 191:80}

if cell_alpha != 90 or cell_beta != 90 or cell_gamma != 90:
    DIPERIODIC_MAPPING.update({1:1, 2:2, 3:3, 6:4, 7:5, 10:6, 13:7})

try:
    group_num_2d = DIPERIODIC_MAPPING[group_num_3d]
except KeyError:
    warnings.warn('No diperiodical group found because rotational axes are inconsistent with 2d translations')
else:
    if   65 <= group_num_2d <= 80: lattice_name = '2d-hexagonal'
    elif 49 <= group_num_2d <= 64: lattice_name = '2d-square'
    elif 8  <= group_num_2d <= 48: lattice_name = '2d-rectangular'
    elif 1  <= group_num_2d <= 7:  lattice_name = '2d-oblique'
```

I hope it can be useful for testing. Taken from [here](https://github.com/spglib/spglib/issues/103#issuecomment-638986086).

**Comment by site-g** [comment]
The layer group implementation now uses a mapping table same as @blokhin mentioned in [#103](https://github.com/spglib/spglib/issues/103#issuecomment-638986086). However, @mikaelkuisma1 gave a reference with different mapping for layer group 35 cm2m in [1306.1280 Table 8](https://arxiv.org/pdf/1306.1280.pdf). I'm not familar with geometric algebra. Could anybody explain the difference?

**Comment by mikaelkuisma1** [comment]
I am continuing the discussion here. Firstly, the code seems excellent, it performed really fast and robustly, at least as far as the sanity check of mapping, back to space group goes (which is both non-injective, non-surjective)).

> How did you run over C2DB 15000 data systematically?

I wrote a preliminary python interface, https://github.com/spglib/spglib/compare/develop...mikaelkuisma1:spglib:add-layer-group-python-interface and a preliminary ASE interface https://gitlab.com/ase/ase/-/merge_requests/2796 .
The script for going through all of the 15k materials can be found here. It takes 1min 40s to run on my laptop.
https://pastebin.com/ddHfaSir . Let me emphasize, that I just did these in a single day, so they are not meant to be close to a final product. I only used it so far to get the layer group number out.

I could make the testing even more systematical, with axis permutations, and orthogonal rotations of cell vectors at least (and maybe even determinant-1 integer matrices).

Is there something else what comes to your mind? I suppose the only thing not tested is that for a given space group of the mapping, there will be a preimage of multiple layer groups, and is the correct one selected.

And ultimately of course also other output data should be tested, not just the layer group number.

> Python, etc. interfaces

If you'd like, addition of doing some benchmarks based on C2DB, this is something I could help with. For yesterday, I basically wanted something to allow me to test things fast, so how it is implemented now, it is just copy-paste from space group classes mostly. However, I would require some consulting from you experts, so that one provides the information in sound sense related to layer groups specifically.

@site-g I agree that the reference I used seems not to be correct. Do you think my fix (Changing from LG35->SG35 to LG35->SG38 is correct)? At least that is what makes the LG code consistent with SG code, as far as I can tell.

I'll have to work on something else today, but I'll get back to you with your tolerance questions later.

**Comment by site-g** [comment]
Python interface is very powerful and convenient, some of my colleagues having been waiting for that for a long time. That's great!

As for other tests, spglib has some artificial structures in its [test folder](https://github.com/spglib/spglib/tree/develop/python/test/data). Maybe we can make some 2D counterparts? But I do not know how they were made. @atztogo
For now I do not have extra idea, other people might have more suggestions?

As for benchmark indicators, hall numbers have more information than spacegroup numbers. Maybe you can try that? I have a mapping table, the blue background color columns in the file map layer hall number to spacegroup hall number. What needs to be noted is that for now the program will output negative numbers to distinguish layer group hall numbers from spacegroup ones.
[Mapping_table.xlsx](https://github.com/spglib/spglib/files/9952874/Mapping_table.xlsx)

LG35->SG38 is what I used, as you can see in the mapping table, I think it's right.

The tolerance question is not urgent, just take your time.

**Comment by mikaelkuisma1** [comment]
I now run the code with 2 million systems (integer det-1, orthogonal rotations and axis permutations of c2db). One system showed tolerance dependence out of 15000.

I added a draft of python interface as a pull request. This is what the python get_layerdataset now returns for MoS2 2H.

Hall number is negative, as mentioned. How that should be returned? How this should look like, are all other names ok, and information relevant for layer groups?
```

{   'choice': '',
    'crystallographic_orbits': array([0, 1, 1], dtype=int32),
    'equivalent_atoms': array([0, 1, 1], dtype=int32),
    'hall': 'p -6 2',
    'hall_number': -114,
    'international': 'p-6m2',
    'mapping_to_primitive': array([0, 1, 2], dtype=int32),
    'number': 78,
    'origin_shift': array([ 0. ,  0. , -0.5]),
    'pointgroup': '-6m2',
    'primitive_lattice': array([[ 3.18      ,  0.        ,  0.        ],
       [-1.59      ,  2.75396078,  0.        ],
       [ 0.        , -0.        , 23.19      ]]),
    'rotations': array([[[ 1,  0,  0],
        [ 0,  1,  0],
        [ 0,  0,  1]],

       [[-1,  1,  0],
        [-1,  0,  0],
        [ 0,  0, -1]],

       [[ 0, -1,  0],
        [ 1, -1,  0],
        [ 0,  0,  1]],

       [[ 1,  0,  0],
        [ 0,  1,  0],
        [ 0,  0, -1]],

       [[-1,  1,  0],
        [-1,  0,  0],
        [ 0,  0,  1]],

       [[ 0, -1,  0],
        [ 1, -1,  0],
        [ 0,  0, -1]],

       [[ 0, -1,  0],
        [-1,  0,  0],
        [ 0,  0, -1]],

       [[ 1,  0,  0],
        [ 1, -1,  0],
        [ 0,  0,  1]],

       [[-1,  1,  0],
        [ 0,  1,  0],
        [ 0,  0, -1]],

       [[ 0, -1,  0],
        [-1,  0,  0],
        [ 0,  0,  1]],

       [[ 1,  0,  0],
        [ 1, -1,  0],
        [ 0,  0, -1]],

       [[-1,  1,  0],
        [ 0,  1,  0],
        [ 0,  0,  1]]], dtype=int32),
    'site_symmetry_symbols': ['-6m2', '3m.', '3m.'],
    'std_lattice': array([[ 3.18      ,  0.        ,  0.        ],
       [-1.59      ,  2.75396078,  0.        ],
       [ 0.        ,  0.        , 23.19      ]]),
    'std_mapping_to_primitive': array([0, 1, 2], dtype=int32),
    'std_positions': array([[ 0.        ,  0.        ,  0.        ],
       [ 0.66666667,  0.33333333,  0.06877965],
       [ 0.66666667,  0.33333333, -0.06877965]]),
    'std_rotation_matrix': array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]]),
    'std_types': array([42, 16, 16], dtype=int32),
    'transformation_matrix': array([[ 1.00000000e+00, -5.55111512e-17,  0.00000000e+00],
       [ 0.00000000e+00,  1.00000000e+00,  0.00000000e+00],
       [ 0.00000000e+00,  0.00000000e+00,  1.00000000e+00]]),
    'translations': array([[0., 0., 0.],
       [0., 0., 1.],
       [0., 0., 0.],
       [0., 0., 1.],
       [0., 0., 0.],
       [0., 0., 1.],
       [0., 0., 1.],
       [0., 0., 0.],
       [0., 0., 1.],
       [0., 0., 0.],
       [0., 0., 1.],
       [0., 0., 0.]]),
    'wyckoffs': ['a', 'f', 'f']}
```

**Comment by site-g** [comment]
The results look rational. As a subgroup of spacegroup, layer group shall contain every information the coresponding spacegroup could given. So all information is relevant to layer groups.

As for the hall number, for human readability it may be split into two entries as something like
```
# for space-group
dataset['group_type'] : 'space_group'
dataset['hall_number']: 481
# for layer group
dataset['group_type']: 'layer_group'
dataset['hall_number']: 114
```
However, sign of the hall number is frequently used for determining the group type, for I was timid in changing a lot of function definition when adding the layer group implementation. For example:
https://github.com/spglib/spglib/blob/876fb9d2499be0fe275018c3000b188c7c6c3a75/src/spg_database.c#L10229-L10243
For now, calling functions like `get_layer_dataset` with the output negative hall number should give the same output, while a positive number will break the result. Besides, modification of `SpglibDataset` might cause compatibility problems, which need more discussion.

So maybe we just leave the negative hall number unchanged as an inside attribute for now, and write some interface functions for for human readability?

**Comment by lan496** [comment]
Adding a new entry like `group_type` seems great to me. One way is to add `SpglibGroupKind` as follows to `SpglibDataset` and `SpglibMagneticDataset`:
```c
typedef enum {
    SPACE_GROUP = 0,
    LAYER_GROUP = 1,
    ROD_GROUP = 2,
    MAGNETIC_SPACE_GROUP = 3,
    MAGNETIC_LAYER_GROUP = 4,
    MAGNETIC_ROD_GROUP = 5,
} SpglibGroupKind;
```
In this case, `set_dataset` and `set_magnetic_dataset` will be good places to add `SpglibGroupKind` and change hall number to be positive. By the way, "type" is a bit overused in Spglib: space-group "type" and "type" of a magnetic space group (e.g. type-IV). So, another key name like `group_kind` will be less confusing.

**Comment by site-g** [comment]
After some search, I found that the terminology [Class](https://en.wikipedia.org/wiki/Class_(set_theory)) and [Class of groups](https://en.wikipedia.org/wiki/Class_of_groups) might meet our requirement

> a class is a collection of [sets](https://en.wikipedia.aufe.cf/wiki/Set_(mathematics)) (or sometimes other mathematical objects) that can be unambiguously defined by a [property](https://en.wikipedia.aufe.cf/wiki/Property_(mathematics)) that all its members share.

So, key names like `group_class` or `class_of_group` could also be included in the candidate list? Will it be confused with crystal class?

Besides, enumeration type seems seldomly used in the code, which I guess is concerned with inconvenience in printing? A string might be enough?

**Comment by lan496** [comment]
@site-g `group_class` is more appropriate!

> Besides, enumeration type seems seldomly used in the code, which I guess is concerned with inconvenience in printing? A string might be enough?

Right. Then, `SpglibDataset` will have a new entry `char group_class[12]` (I am not sure the length is enough, but sufficient for at least "space_group", "layer_group", and "rod_group").

**Comment by mikaelkuisma1** [comment]
There are lots of good suggestions. I would be happy to help implementing them as a draft, if you feel that they are ready for that.

@atztogo @site-g We would like to contact you via e-mail. @site-g I was not able to find your e-mail, could you send me a message to kuisma@dtu.dk .

**Comment by atztogo** [comment]
> @site-g `group_class` is more appropriate!
> 
> > Besides, enumeration type seems seldomly used in the code, which I guess is concerned with inconvenience in printing? A string might be enough?
> 
> Right. Then, `SpglibDataset` will have a new entry `char group_class[12]` (I am not sure the length is enough, but sufficient for at least "space_group", "layer_group", and "rod_group").

I think enum is cleaner at least in C.

**Comment by atztogo** [comment]
> As for other tests, spglib has some artificial structures in its [test folder](https://github.com/spglib/spglib/tree/develop/python/test/data). Maybe we can make some 2D counterparts? But I do not know how they were made. @atztogo For now I do not have extra idea, other people might have more suggestions?

The artificial structures were generated as a by-product of cluster-expansion study by my former colleague. It starts with high symmetry crystal structure and some atoms are replaced. Lattice is kept same, but some space group operations are broken due to the change of atomic arrangement.

**Comment by atztogo** [comment]
> So maybe we just leave the negative hall number unchanged as an inside attribute for now, and write some interface functions for for human readability?

Use of negative number is implicit. So it would be nice to find a good way to avoid it in the future. Efficiency for `grep` and `diff` or git is also important. This is just a hope. Easy-to-say, but difficult-to-do. I'm sorry that currently I have no concrete idea.

**Comment by site-g** [comment]
> I think enum is cleaner at least in C.

Then we may use enum. I suggest string just because there is no enum in `SpglibDataset` at present.

> Use of negative number is implicit. So it would be nice to find a good way to avoid it in the future. Efficiency for grep and diff or git is also important.

That is right. To make it clear, one way might be passing the not yet implemented `group_class` into inside codes, but then huge amount of declarations will change. Another way might be making `hall_number` an enum, but the relationship between int and enum in C seems [implementation-defined](https://stackoverflow.com/a/63249952). I am not experienced at coding and do not have more elgant scheme.

**Comment by askhl** [comment]
Hi everyone!

It is very nice to have the layergroup code.  However currently it resides in `spglib._internal`, so that's not meant to be the permanent.  Can we decide where the code should go?

Another issue is that some of the spglib functions are available only for 3D systems.  The only functions that know about layer groups are get_symmetry_layerdataset and get_layergroup.  There are also also some in `spglib.cell`: `standardize_cell()`, `refine_cell()`, and `find_primitive()`.  Those functions always call the 3D versions of the API.

However I was thinking: Maybe there should be versions of find_primitive() etc. which do *not* take symprec, but instead take an existing dataset.   This means the dataset is only established initially once and for all, and then one gets access to find_primitive() no matter how the dataset was established.

It would also be nice to know if there's any other work remaining for the layer group code in order to be considered 'done'.

**Comment by atztogo** [comment]
I don't think the current implementation is logically clean with respect to the definition of a layer group, since it is not based on the direct definition of a layer group as the stabilizer of a layer in a 3D space group. This could lead to inconsistencies between non-specialist users' understanding of the implemented layer group and its group-theoretical definition. Given the potential impact on the community if a critical deficiency is found later — which might force us to introduce an inelegant local rule — I'd prefer not to support the current implementation.

If we want to support layer groups properly, we should discard the current implementation and rebuild it from scratch at the layer-group level. And if we go that route in the future, I think it should be implemented in moyo rather than in spglib.

What do you think @lan496?

**Comment by lan496** [comment]
I have not implemented layer-group support in Moyo yet, simply because I have not had a concrete use case for it in my own research so far. That said, I agree that it is a very interesting direction, and I would like to think about how the layer-group case could fit into Moyo’s current codebase .

Regarding the algorithm, I also think there is room for improvement in terms of simplicity and clarity, especially if we want the implementation to stay close to the mathematical definition.

**Comment by atztogo** [comment]
Thanks @lan496, but please don't feel obligated. @askhl, I no longer plan to support layer groups within the current spglib implementation. Since this issue is misleading, I'd like to close it.

**Comment by askhl** [comment]
In spite of these problems, we are using the layer group code.  To continue doing so, do we have to fork spglib or is there some alternative?

**Comment by LecrisUT** [comment]
I would urge an alternative to forking spglib. @askhl, can you open a new issue with more explicit details on what would be the minimal interface that you would need exposed, and we can at least work on exposing that. Probably not as a complete public interface, unless the python interface can be made compatible with moyo's.

@atztogo @lan496 do note that an interface layer would still be needed for spglib -> moyo transition. The sooner a high level interface (mostly python or any other OOP language) can be fixed the better because then both spglib and moyo's interface can converge together on it.

**Comment by askhl** [comment]
I have opened #631 now.

Thanks everyone.

**Comment by atztogo** [comment]
@LecrisUT, I don't want to carry over spglib's poor theoretical or methodological choices into moyo. In other words, a complete transition is simply not possible.

**Comment by LecrisUT** [comment]
> [@LecrisUT](https://github.com/LecrisUT), I don't want to carry over spglib's poor theoretical or methodological choices into moyo. In other words, a complete transition is simply not possible.

I don't believe that is the request here. The request is just for high level API to be available. Internally it can do completely different operations.

Some API change is fine as long as there is a transition path to achieve the overall workflow. @askhl shared one such workflow in #631 and additional eyes on that and if it overall compatible or if there would be better alternatives would be good.

Basically my suggestion is to collect the user stories of what they are trying to do with this part and consider those in the moyo approach to it, even if it ends up being a documentation note.

**Comment by atztogo** [comment]
The difficulty is that this isn't purely a software implementation problem, it also involves crystallography and the so-called "2D" concept, whose definition I personally find questionable.

**Comment by LecrisUT** [comment]
> The difficulty is that this isn't purely a software implementation problem, it also involves crystallography and the so-called "2D" concept, whose definition I personally find questionable.

Sure but isn't that because they are trying to fit a square cube in a round hole, i.e. finding the 2D (heck why not even 1D) periodic unit with tools that have a fixed 3D representation? If the question/API is changed to, given a collection of points in a 2D surface, can we find a periodic unit cell, its translation and rotation symmetries, then yes, I do not see a physical or mathematical flaw in that.

And that's why I have the recommendation to gather the user stories, what exactly are they trying to do. Implementing the same logic that spglib used and be bug-to-bug equivalent, no that is not the goal here. But having an answer to the user's question of "I was doing this, and now I cannot do it anymore because spglib 2.x does not compile on python 4.0, what do I do?", that is still something that should be answered. The answer could be that spglib was not doing something correct there, but that does not imply that the user was also trying to do something wrong, it just may look like it if they forcing the square cube into the round hole because the cilinder was not available to them.

**Comment by atztogo** [comment]
It seems we might not be fully aligned on each other's thinking just yet. If you don't mind, I'd like to reach out to @askhl directly to help move things along more efficiently.

**Comment by LecrisUT** [comment]
> It seems we might not be fully aligned on each other's thinking just yet. If you don't mind, I'd like to reach out to [@askhl](https://github.com/askhl) directly to help move things along more efficiently.

No, we are aligned there. That is basically my suggestion in much less words 😉

**Comment by askhl** [comment]
Thanks for the discussion.  I'll elaborate our use case a little bit more.

We use spglib's layergroup code to classify materials for the Computational 2d materials database, C2DB (https://c2db.fysik.dtu.dk/).  Each material has a layer group listed, just like one might list a spacegroup for 3D crystals.  It would be surprising *not* to make an effort to show a material's symmetry.  For example https://c2db.fysik.dtu.dk/material/2Bi-1 says this material has layer group 72.  For consistency, aside from just listing the symmetries, we are also using them in computations.

The 2D materials have an ordinary 3D structure for a basis, and they repeat on a 2D lattice in the xy plane and not in the z direction.  I cannot see that there is anything fundamentally questionable or physically or mathematically ill-defined about characterizing symmetries in such systems.  A library like spglib, which contains all the requisite functionality for symmetry determination in 3D, is a very useful for solving also the problem for 2D materials.  Hence the implementation.

But please feel free to elaborate if there is some other fundamental problem.
